### PR TITLE
feat: legalization from xten_nn.reduce_mean to aten.mean.dim

### PR DESCRIPTION
**Pull request by ehsan-toosi**

### CODE REVIEW

File: lib/Conversion/XTenNNToTorch.cpp [hunk 1]
```diff
@@ -220,6 +220,21 @@ convTranspose2dToTorch(ConvTransposeOp op, ConvTransposeOp::Adaptor adaptor,
...
+                  ArrayRef<Type> types, ValueRange values,
+                  ConversionPatternRewriter &rewriter) {
+  auto loc = op->getLoc();
+  auto noneConst = rewriter.create<Torch::ConstantNoneOp>(loc);
+  auto keepdims =
+      rewriter.create<Torch::ConstantBoolOp>(loc, adaptor.getKeepdims());
+  auto axes = Torch::toTorchList(loc, rewriter, adaptor.getAxes().vec());
+  return rewriter
+      .create<Torch::AtenMeanDimOp>(loc, types[0], values[0], axes, keepdims,
+                                    noneConst)
```
Comment (p-lanza) [inline lib/Conversion/XTenNNToTorch.cpp:234]: Out of ignorance, what is the `noneConst` for?

**Review by p-lanza — COMMENTED** [review]
LGTM though admittedly I'm not familiar with `torch-mlir`. We could potentially make it into a generic function as well if we want to support more reduce operators.

File: lib/Conversion/XTenNNToTorch.cpp (hunk 1 above)
Comment (ehsan-toosi) [reply lib/Conversion/XTenNNToTorch.cpp:234]: It's for dtype: https://pytorch.org/docs/stable/generated/torch.mean.html

**Review by roberteg16 — APPROVED** [review]
LGTM, thanx

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp [hunk 2]
```diff
@@ -593,3 +597,39 @@ bool TopK::isCompatibleReturnTypes(mlir::TypeRange l, mlir::TypeRange r) {
       getElementTypeOrSelf(l[1]) == getElementTypeOrSelf(r[1]);
   return sameElementType && succeeded(verifyCompatibleShapes(l, r));
 }
+
+LogicalResult ReduceMeanOp::inferReturnTypeComponents(
+    MLIRContext * /*context*/, std::optional<Location> location,
+    ReduceMeanOp::Adaptor adaptor,
+    SmallVectorImpl<ShapedTypeComponents> &inferredReturnShapes) {
```
Comment (ehsan-toosi) [inline lib/Dialect/XTenNN/IR/XTenNNOps.cpp:604]: Is there a test for it @p-lanza ?

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp [hunk 3]
```diff
@@ -593,3 +597,39 @@ bool TopK::isCompatibleReturnTypes(mlir::TypeRange l, mlir::TypeRange r) {
...
+    // normalize axis: [0, r)
+    if (axis < 0) {
+      axis += inTy.getRank();
+    }
+    assert((axis >= 0 && axis < inTy.getRank()) && "axis has invalid value");
+
+    if (keepDims) {
+      newAxes[axis] = 1;
+    } else {
+      newAxes.erase(newAxes.begin() + axis);
```
Comment (ehsan-toosi) [inline lib/Dialect/XTenNN/IR/XTenNNOps.cpp]: This is going to go out of bound in the next iteration because you remove from the shape in the previous iteration of axes. @p-lanza

Comment (ehsan-toosi) [reply lib/Dialect/XTenNN/IR/XTenNNOps.cpp]: I think it'd be better to go over the dims of input and if it's among the axis, push back to `inDims` only if keepdims through.

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp (hunk 2 above)
Comment (ehsan-toosi) [reply lib/Dialect/XTenNN/IR/XTenNNOps.cpp:604]: I see. It's being checked at reduce_mean.mlir

File: lib/Dialect/XTenNN/IR/XTenNNOps.cpp (hunk 3 above)
Comment (ehsan-toosi) [reply lib/Dialect/XTenNN/IR/XTenNNOps.cpp]: Fixed it.

**Review by p-lanza — APPROVED** [review]
